_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,332 @@
+- Then if we have one, we fetch the name of the event's user,
+- And we map the future result of this task to the `Name` message.
+
+But now, how are we going to run all this and execute the actual HTTP requests and process the responses accordingly? We need to setup the [Elm Architecture]:
+
+```haskell
+module Main exposing (main)
+
+import Html exposing (..)
+import Http
+import Json.Decode as Decode
+import Task exposing (Task)
+
+
+type alias Model =
+    { name : Maybe String
+    , error : String
+    }
+
+type Msg
+    = Name (Result String String)
+
+eventsRequest : Http.Request (List String)
+eventsRequest =
+    Http.get "https://api.github.com/events"
+        (Decode.at [ "actor", "login" ] Decode.string
+            |> Decode.list
+        )
+
+nameRequest : String -> Http.Request String
+nameRequest login =
+    Http.get ("https://api.github.com/users/" ++ login)
+        (Decode.at [ "name" ]
+            (Decode.oneOf
+                [ Decode.string
+                , Decode.null "unspecified"
+                ]
+            )
+        )
+
+toHttpTask : Http.Request a -> Task String a
+toHttpTask request =
+    request
+        |> Http.toTask
+        |> Task.mapError toString
+
+fetchEvents : Task String (List String)
+fetchEvents =
+    toHttpTask eventsRequest
+
+fetchName : String -> Task String String
+fetchName login =
+    toHttpTask (nameRequest login)
+
+pickFirst : List String -> Task String String
+pickFirst events =
+    case List.head events of
+        Just event ->
+            Task.succeed event
+
+        Nothing ->
+            Task.fail "No events."
+
+init : ( Model, Cmd Msg )
+init =
+    { name = Nothing, error = "" }
+        ! [ fetchEvents
+                |> Task.andThen pickFirst
+                |> Task.andThen fetchName
+                |> Task.attempt Name
+          ]
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        Name (Ok name) ->
+            { model | name = Just name } ! []
+
+        Name (Err error) ->
+            { model | error = error } ! []
+
+view : Model -> Html Msg
+view model =
+    div []
+        [ if model.error /= "" then
+            div []
+                [ h4 [] [ text "Error encountered" ]
+                , pre [] [ text model.error ]
+                ]
+          else
+            text ""
+        , p [] [ text <| Maybe.withDefault "Fetching..." model.name ]
+        ]
+
+main =
+    Html.program
+        { init = init
+        , update = update
+        , subscriptions = always Sub.none
+        , view = view
+        }
+```
+
+That's for sure much more code than with the JavaScript example, but don't forget that the Elm version renders HTML, not just logs in the console, and that the JavaScript code could be refactored to look a lot like the Elm version.
+
+As always, an [Ellie](https://ellie-app.com/7Q9svdqRGa1/2) is publicly available so you can play around with the code.
```

Comment:
It seems that the Ellie is not up to date, I didn't make a diff, but there is at least a difference on line 303: it misses the "Fetching..." default value in the Ellie version. :)